README.md
# accessible_text_view

Renders a native platform text view to address accessibility with embedded links in Flutter Text and RichText widgets on Android.

Also works on iOS, with enhanced accessibility options.

## Accessibility Problems Solved

### Android

Flutter Text and Text.rich() or RichText widgets cannot currently handle links properly, until the Google Android team exposes the "link trait" in its `AccessibilityNodeInfo` class. Currently, only `URLSpan` spans in `SpannableString` instances on Android expose their links properly to TalkBack.

The expected behavior is for the entire text to be read, with a slight pause and distinctive "popping click" sound heard at each link in the text. The user can then tap with three fingers to get the TalkBack menu. In this context menu, there is a Links submenu containing the links. This behavior cannot currently be achieved by the Flutter framework because of the lack of public Android API available for this functionality.

__However, I also wish to advocate here for an option in Android to make these elements separately focusable for longer paragraphs and blocks of text.__ This would help TalkBack users to not have to listen to the entire text. The current Flutter Text widget implementation does support the links as separate focus nodes, but calls them out as buttons instead of links. If the Android team makes the "link trait" available in its public API so that the Flutter engine can use it without having to use a `TextView` with `Spannable` text, this end could be easily achieved by allowing `Semantics(link: true)` to work on Flutter `Text`, `RichText`, and `TextSpan`s.

### iOS

Fortunately, `UIKit` has an `UIAccessibilityTraits.link` attribute, and thus the Flutter links within `Text` and `TextSpan`s are correctly labelled. Additionally, Flutter gets it right by making them separate focus nodes by default. Unfortunately, iOS users may still get confused, because Apple has changed `UITextView` to behave more like Android `TextView`, in that links in a `UITextView` are no longer separate focus nodes. VoiceOver users must be more advanced and know to use the links rotor to access them. It seems that screen reader users are still left largely in the dark when it comes to embedded links in text, an all too common phenomenon in native apps, e.g. terms of service, support contact info, usage instructions, etc.

To brige the gap, I have enhanced the `UITextView` accessibility behavior by allowing the developer to choose one of the following beaviors appropriate to their app and the amount of text in the text view:


You can also use platform tests to use a Flutter implementation on iOS, and use this widget on Android, but I implemented native views on both platforms to facilitate easier implementation.

Besides, you can simply embed your links in HTML, and use data auto-detection (both Android and iOS) to auto-link phone numbers and e-mail addresses!

## Installation

`flutter pub add accessible_text_view`

### Dependencies

Other than Flutter, none!

## Usage Example

```dart
class HomePage extends StatefulWidget {
  const HomePage({super.key});

  @override
  State<HomePage> createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  int currentTabIndex = 0;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
    appBar: AppBar(
        title: Semantics(
            header: true,
            child: const Text('Flutter TextView example'),
        ),
    ),
    backgroundColor: Colors.black38,
    body: Column(
      mainAxisAlignment: MainAxisAlignment.start,
      children: [
        Padding(
          padding: const EdgeInsets.all(8.0),
          child: AccessibleTextView(
            html: """
            <p>This is an AccessibleTextView widget. Its behavior is more correct on Android and more flexible on iOS.</p>
            <p>This is a second paragraph.</p>
            <p>This is an example text with phone number 555-234-2345 with an e-mail link of test@example.com and an external link to <a href="https://google.com">Google</a>.
            On Android, TalkBack should indicate each link with a blip sound.
            On iOS, VoiceOver should provide a long-press links menu.</p>
            <p>This component really should not be used for long text, because it is not designed for long text. For that, use a web view.</p>
          """,
              textColor: Colors.white,
              linkWeight: FontWeight.bold,
              linkColor: Colors.blueAccent,
              appearance: AccessibleTextViewAppearance.dark,
              fontSize: 14,
            ),
          ),
        ],
      ),
    );
  }
}
```

## License

This plugin has a liberal MIT license to encourage accessibility in all app development. Please learn from it and use as you see fit in your own apps!

## Contributing

I would really appreciate learning who is using this plugin, and your feedback, and bugfix and feature requests. Please feel free to open an issue or pull request.

## Contributors

- Tom Grushka, principal developer
- Adam Campfield


CHANGELOG.md
## 1.1.0

* Breaking change: now uses `TextStyle` for text and link styles.
* Fix line height computation on Android.
* Allow Flutter font text scaling behavior on iOS.
* Graceful fallback to system font sizes when `fontSize` is not provided.
* Demonstrates `native_segments` accessible segmented control plugin in the example.

## 1.0.3

* Bump native_flutter_fonts version.

## 1.0.2

* Bump native_flutter_fonts version.

## 1.0.1

* Get working with native_flutter_fonts on iOS and Android.

## 1.0.0

* Initial release.


example/lib/main.dart
import 'package:accessible_text_view/accessible_text_view.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:native_segments/native_segments.dart';
import 'package:url_launcher/url_launcher_string.dart';

void main() => runApp(const MaterialApp(home: TextViewExample()));

class TextViewExample extends StatefulWidget {
  const TextViewExample({super.key});

  @override
  State<TextViewExample> createState() => _TextViewExampleState();
}

class _TextViewExampleState extends State<TextViewExample> {
  bool isFlutterTextView = false;

  Widget get accessibleTextView => Padding(
        padding: const EdgeInsets.only(top: 10),
        child: AccessibleTextView(
          html: html,
          textStyle: normalStyle,
          linkStyle: linkStyle,
          brightness: Brightness.dark,
          useFlutterTextScale: true,
        ),
      );

  final String html1 = """
  This is short text.
""";

  final String html = """
    <p>This is an AccessibleTextView.</p>
    <p>This is a second paragraph with one <a href="https://google.com">Google</a> link.</p>
    <p>This is an example text with phone number 555-234-2345 with an e-mail link of test@example.com and a web link to <a href="https://google.com">Google</a>.
    On Android, TalkBack should indicate each link with a blip sound.
    On iOS, VoiceOver should navigate each paragraph separately, and provide a rotor and long-press popup menu (in VoiceOver only) for links.</p>
    <p>This component really should not be used for long text, because it is not designed for long text. For that, use a web view.</p>
  """;

  final TextStyle normalStyle = const TextStyle(
    color: Colors.white,
    fontSize: 14,
    height: 1.5,
  );

  final TextStyle linkStyle = const TextStyle(
    color: Colors.blueAccent,
    fontWeight: FontWeight.bold,
    decoration: TextDecoration.underline,
  );

  late final Widget flutterTextView = Text.rich(
    TextSpan(
      text: 'This is a Flutter Text.rich.',
      children: [
        const TextSpan(text: '\n\nThis is a second paragraph with one '),
        TextSpan(
          text: 'Google',
          style: linkStyle,
          recognizer: TapGestureRecognizer()
            ..onTap = () async {
              await launchUrlString('https://google.com');
            },
        ),
        const TextSpan(text: ' link.'),
        TextSpan(
          text: '\n\nThis is an example text with phone number ',
          children: [
            TextSpan(
              text: '555-234-2345',
              style: linkStyle,
              recognizer: TapGestureRecognizer()
                ..onTap = () async {
                  await launchUrlString('tel:555-234-2345');
                },
            ),
            const TextSpan(text: ' with an e-mail link of '),
            TextSpan(
              text: 'test@example.com',
              style: linkStyle,
              recognizer: TapGestureRecognizer()
                ..onTap = () async {
                  await launchUrlString('mailto:test@example.com');
                },
            ),
            const TextSpan(text: ' and an external link to '),
            TextSpan(
              text: 'Google',
              style: linkStyle,
              recognizer: TapGestureRecognizer()
                ..onTap = () async {
                  await launchUrlString('https://google.com');
                },
            ),
            const TextSpan(
                text:
                    '. On Android, TalkBack should indicate each link with a blip sound. '),
            const TextSpan(
                text:
                    'On iOS, VoiceOver should navigate each non-link and link chunk separately.'),
          ],
        ),
        const TextSpan(
            text:
                '\n\nThis component really should not be used for long text, because it is not designed for long text. For that, use a web view.'),
      ],
    ),
    style: normalStyle,
  );

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: Semantics(
            header: true,
            child: const Text('Flutter TextView example'),
          ),
        ),
        backgroundColor: Colors.black38,
        body: Column(
          children: [
            Padding(
              padding: const EdgeInsets.all(8.0),
              child: [TargetPlatform.iOS, TargetPlatform.android]
                      .contains(defaultTargetPlatform)
                  ? NativeSegments(
                      segments: [
                        NativeSegment(title: 'AccessibleTextView'),
                        NativeSegment(title: 'Flutter Text.rich'),
                      ],
                      style: const NativeSegmentsStyle(
                        isDarkTheme: true,
                      ),
                      onValueChanged: (value) {
                        setState(() {
                          isFlutterTextView = value == 1;
                        });
                      },
                    )
                  : SegmentedButton(
                      emptySelectionAllowed: false,
                      multiSelectionEnabled: false,
                      segments: const [
                        ButtonSegment(
                          value: 0,
                          label: Text('AccessibleTextView'),
                        ),
                        ButtonSegment(
                          value: 1,
                          label: Text('Flutter Text.rich'),
                        ),
                      ],
                      selected: {isFlutterTextView ? 1 : 0},
                      onSelectionChanged: (selectedSet) {
                        setState(() {
                          isFlutterTextView = selectedSet.contains(1);
                        });
                      },
                    ),
            ),
            Expanded(
              flex: 3,
              child: isFlutterTextView ? flutterTextView : accessibleTextView,
            ),
            Expanded(
                flex: 1,
                child: Container(
                    color: Colors.blue[100],
                    child: const Center(child: Text("Hello from Flutter!"))))
          ],
        ),
      ),
    );
  }
}
